horovod/common/operations.cc

```diff
@@ -840,18 +847,147 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
       if (i == 0) {
         displcmnts[i] = 0;
       } else {
-        displcmnts[i] = recvcounts[i - 1] + displcmnts[i - 1];
+        displcmnts[i] = displcmnts[i - 1] + recvcounts[i - 1];
       }
     }
-    auto result = MPI_Allgatherv(
-        e.tensor->data(), (int)e.tensor->shape().num_elements(),
-        GetMPIDataType(e.tensor), (void*)e.output->data(), recvcounts,
-        displcmnts, GetMPIDataType(e.tensor), horovod_global.mpi_comm);
+
+    // If the cluster is homogeneous do parallelized MPI cross-node
+    // allgather into the shared buffer first, then memcpy to output buffer
+    if (horovod_global.is_homogeneous) {
+      int element_size;
+      MPI_Type_size(GetMPIDataType(e.tensor), &element_size);
+
+      auto* cross_recvcounts = new int[horovod_global.cross_size]();
+      auto* cross_displcmnts = new int[horovod_global.cross_size]();
+      for (int i = 0; i < horovod_global.cross_size; i++) {
+        cross_recvcounts[i] = recvcounts[horovod_global.local_size * i +
+                                         horovod_global.local_rank];
+        cross_displcmnts[i] = displcmnts[horovod_global.local_size * i +
+                                         horovod_global.local_rank];
+      }
+      int64_t total_size = recvcounts[tensor_sizes.size() - 1] +
+                           displcmnts[tensor_sizes.size() - 1];
+
+      // If shared buffer is not initialized or is not large enough, reallocate
+      if (horovod_global.shared_buffer == nullptr ||
+          horovod_global.shared_buffer_size < total_size * element_size) {
+        if (horovod_global.shared_buffer != nullptr) {
+          MPI_Win_fence(0, horovod_global.window);
+          MPI_Win_free(&horovod_global.window);
+          horovod_global.shared_buffer = nullptr;
+        }
+        int64_t window_size =
+            horovod_global.local_rank == 0 ? total_size * element_size : 0;
+
+        // Allocate shared memory, give each rank their respective pointer
+        ACTIVITY_START_ALL(entries, timeline, ALLOCATE_SHARED_BUFFER)
+        MPI_Win_allocate_shared(
+            window_size, element_size, MPI_INFO_NULL, horovod_global.local_comm,
+            &horovod_global.shared_buffer, &horovod_global.window);
+
+        if (horovod_global.local_rank != 0) {
+          int disp_unit;
+          MPI_Aint winsize;
+          MPI_Win_shared_query(horovod_global.window, 0, &winsize, &disp_unit,
+                               &horovod_global.shared_buffer);
+        }
+        MPI_Win_fence(0, horovod_global.window);
```

Comment:
This is the first time I'm seeing MPI Windows, but from the cursory reading of the docs, you need a fence for MPI_Get/MPI_Put operations.  Do you need a fence for operating on (shared) buffers with other operations, such as MPI_Allgather?

Reply:
I had the fences before/after `MPI_Allgatherv` just to be safe, but since Allgather is blocking/synchronous, it seems this is not needed (my tests worked fine without them). I kept the one during the initialization of the shared memory buffer because (i) it does not happen regularly so the cost of synchronization is negligible, (ii) we do not want one rank to call `MPI_Win_free` while the other is using it.